programs/sunrise_beam/src/instructions/mint_gsol.rs

```diff
@@ -1,61 +1,57 @@
-use crate::{
-    seeds::GSOL_MINT_AUTHORITY, state::ControllerState, token, utils::get_cpi_program_id,
-    BeamProgramError,
-};
 use anchor_lang::prelude::*;
-use anchor_lang::solana_program::sysvar::{
-    instructions::Instructions as SysvarInstructions, SysvarId,
-};
-use anchor_spl::token::{Mint, Token, TokenAccount};
-
-#[derive(Accounts)]
-pub struct MintGsol<'info> {
-    #[account(
-        mut,
-        has_one = gsol_mint,
-    )]
-    pub state: Box<Account<'info, ControllerState>>,
-    pub beam: Signer<'info>,
-    #[account(mut)]
-    pub gsol_mint: Box<Account<'info, Mint>>,
-    #[account(
-        seeds = [state.key().as_ref(), GSOL_MINT_AUTHORITY],
-        bump = state.gsol_mint_authority_bump
-    )]
-    pub gsol_mint_authority: SystemAccount<'info>,
-    #[account(mut, token::mint = gsol_mint)]
-    pub mint_gsol_to: Box<Account<'info, TokenAccount>>,
-    pub token_program: Program<'info, Token>,
-    /// CHECK: We check that the address is that of the instructions sysvar.
-    #[account(constraint = SysvarInstructions::check_id(sysvar.key))]
-    pub sysvar: UncheckedAccount<'info>,
-}
+
+use crate::system::check_beam_validity;
+use crate::utils::get_cpi_program_id;
+use crate::{token, BeamError, MintGsol};
 
 pub fn handler(ctx: Context<MintGsol>, amount: u64) -> Result<()> {
-    let accounts = &ctx.accounts;
+    let state = &mut ctx.accounts.state;
+    let gsol_mint = &ctx.accounts.gsol_mint;
+
+    let cpi_program = get_cpi_program_id(&ctx.accounts.sysvar.to_account_info())?;
+    check_beam_validity(state, &ctx.accounts.beam, &cpi_program)?;
+
+    let can_mint = {
+        let pre_supply = state.pre_supply;
+        let details = state.get_mut_beam_details(&ctx.accounts.beam.key());
+        if details.is_none() {
+            return Err(BeamError::UnidentifiedBeam.into());
+        }
+
+        let allocation = details.unwrap();
+        let effective_supply = gsol_mint.supply.checked_sub(pre_supply).unwrap();
 
-    let cpi_program = get_cpi_program_id(&accounts.sysvar.to_account_info())?;
-    crate::check_beam_parameters(
-        &accounts.state,
-        &accounts.beam.to_account_info(),
-        &cpi_program,
-    )?;
+        let mint_window = if effective_supply != 0 {
+            (allocation.allocation as u64)
+                .checked_mul(effective_supply)
+                .unwrap()
+                .checked_div(100)
+                .unwrap()
+        } else {
+            // Let the first request mint without restrictions. The allocations will come into effect afterwards.
```

Comment:
I think the next step is to add tests around this logic. Both internal unit tests and functional tests.

Reply:
Yh that'll be coming up next